### PR TITLE
Add `PathBuf::as_mut_os_string` and `Path::as_mut_os_str`

### PR DESCRIPTION
**Pull request by zertosh**
Implements rust-lang/libs-team#140 (tracking issue #105021).

### CODE REVIEW

**Comment by zertosh** [comment]
Not sure if this is worth marking as "unstable", but if it is, not sure what to use for "issue" since those seem to point to `rust-lang/rust` issues, and not `rust-lang/libs-team` issues.

And if it's not worth marking "unstable", not sure what to put for `since`.

**Comment by ChrisDenton** [comment]
Using `none` is fine initially but this feature should have [a tracking issue](https://github.com/rust-lang/rust/issues/new?assignees=&labels=C-tracking-issue%2C+T-libs-api&template=library_tracking_issue.md&title=Tracking+Issue+for+XXX) so that stabilization progress can be tracked. Once that's created, the "issue" number can refer to that.

Also note that adding a `DerefMut` implementation is public and instantly stable so would require an FCP (a vote on stabilizing it) before it can be merged. It would likely be better to implement this without it and then add it separately if desired.

File: library/std/src/path.rs
Comment (dtolnay) [inline library/std/src/path.rs:1756]: Please move this impl to a separate PR because there is no mechanism to feature-gate DerefMut impls, so it requires team approval to merge as instantly stable.

Does there end up being no possible way to write example code for `as_mut_os_str` without this?

You could try `into_boxed_path` (with a comment that links to the DerefMut PR, and a note in the tracking issue to follow up with more realistic example code after DerefMut is available).

File: library/std/src/path.rs
Comment (dtolnay) [inline library/std/src/path.rs:1750]:
```suggestion
#[stable(feature = "path_buf_deref_mut", since = "1.67.0")]
```

File: library/std/src/path.rs
Comment (dtolnay) [inline library/std/src/path.rs:2047]: After filing a library tracking issue in rust-lang/rust, these attributes need to get updated with an issue number.

**Review by dtolnay — CHANGES_REQUESTED** [review]
Thanks!

**Review by dtolnay — APPROVED** [review]

**Pull request merged**
